### PR TITLE
[WIP] GlusterFS volume driver

### PR DESCRIPTION
**Pull request by jarrpa**
This PR implements a GlusterFS CSI volume driver. It is based on the Kubernetes in-tree GlusterFS volume plugin.

**WARNING:** This is a WIP PR, currently intended as a place to gather feedback while I figure out what the heck I'm doing. :)

### CODE REVIEW

File: pkg/glusterfs/driver.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+func NewDriver(nodeID, endpoint string) *driver {
+	glog.Infof("Driver: %v version: %v", driverName, csicommon.GetVersionString(&version))
+
+	d := &driver{}
+
+	d.endpoint = endpoint
+
+	csiDriver := csicommon.NewCSIDriver(driverName, &version, GetSupportedVersions(), nodeID)
+	csiDriver.AddControllerServiceCapabilities([]csi.ControllerServiceCapability_RPC_Type{csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME})
+	csiDriver.AddControllerServiceCapabilities([]csi.ControllerServiceCapability_RPC_Type{csi.ControllerServiceCapability_RPC_PUBLISH_UNPUBLISH_VOLUME})
```
Comment (rootfs) [inline pkg/glusterfs/driver.go]: co publish/unpublish are empty

File: app/glusterfsplugin/main.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+func main() {
+	cmd := &cobra.Command{
+		Use:   "GlusterFS",
+		Short: "GlusterFS CSI driver",
+		Run: func(cmd *cobra.Command, args []string) {
+			handle()
+		},
+	}
+
+	cmd.PersistentFlags().StringVar(&mode, "mode", "", "operating mode: 'node' or 'controller'")
```
Comment (rootfs) [inline app/glusterfsplugin/main.go]: why dual mode?

**Comment by lpabon** [comment]
@jarrpa Take a look at https://github.com/kubernetes-csi/docs/wiki/Drivers

**Comment by lpabon** [comment]
Also notice the driver pattern in https://github.com/kubernetes-csi/docs/wiki/Drivers and in https://github.com/kubernetes-csi/csi-test/blob/master/driver/driver.go

**Comment by lpabon** [comment]
Lastly, it would be great for you to read https://github.com/kubernetes-csi/docs/wiki/Development and provide feedback back to me.

File: app/glusterfsplugin/main.go (hunk 2 above)
Comment (jarrpa) [reply app/glusterfsplugin/main.go]: I think I remember seeing it as a suggested pattern somewhere in the CSI docs. If it doesn't really matter I can just do a unified Serve call.

File: pkg/glusterfs/driver.go (hunk 1 above)
Comment (jarrpa) [reply pkg/glusterfs/driver.go]: As in, my functions are empty, or the capabilities don't do anything?

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @jarrpa isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Pull request closed without merging**